pkg/ebpf/map.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpf
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/lock"
+
+	ciliumebpf "github.com/cilium/ebpf"
+)
+
+type MapSpec = ciliumebpf.MapSpec
+
+const (
+	PerCPUHash = ciliumebpf.PerCPUHash
+
+	PinByName = ciliumebpf.PinByName
+)
+
+// IterateCallback represents the signature of the callback function expected by
+// the IterateWithCallback method, which in turn is used to iterate all the
+// keys/values of a map.
+type IterateCallback func(key, value interface{})
+
+// Map represents an eBPF map.
+type Map struct {
+	*ciliumebpf.Map
+
+	spec *MapSpec
+
+	path string
+	lock lock.RWMutex
+}
+
+// NewMap creates a new Map object.
+func NewMap(spec *MapSpec) *Map {
+	return &Map{
+		spec: spec,
+	}
+}
+
+// OpenOrCreate tries to open or create the eBPF map identified by the spec in
+// the Map object.
+func (m *Map) OpenOrCreate() error {
```

Comment:
This function does more than open/create, it pins it as well, which is not obvious from the fn name or docstring.

Since it doesn't take any arguments, wouldn't this code be better suited to live in `NewMap()` instead? Is there a requirement for this not to be tied to instantiation of the map?

For example, this can be added to `NewMap()` including a `pin bool` argument to control the pinning behaviour.

Reply:
> This function does more than open/create, it pins it as well which is not obvious from the fn name or docstring.

This function pins only if `MapSpec.Pinning` (passed to `ebpf.NewMap()`) is set to `ebpf.PinByName` (see for example https://github.com/cilium/cilium/blob/pr/jibi/move-metrics-to-cilium-ebpf/pkg/maps/metricsmap/metricsmap.go#L199)

> Since it doesn't take any arguments, wouldn't this code be better suited to live in NewMap() instead? Is there a requirement for this not to be tied to instantiation of the map?

I wanted to keep the current behavior as I'm not sure yet if there are occasions where it's actually useful to separate the map "definition" phase from the instantiation phase